### ISSUE TITLE
Allow overwriting settings from log4j.properties with environment properties

### ISSUE THREAD

**Issue by georgms** [getsentry/raven-java#266]
Currently in `raven-log4j` the settings from `log4j.properties` have the highest priority of all three ways of configuration (env vars, system properties, `log4j.properties`):
- `com.getsentry.raven.jul.SentryHandler#SentryHandler()` fetches settings from env vars and system properties and then calls ...
- `com.getsentry.raven.jul.SentryHandler#retrieveProperties` which fetches settings from `log4j.properties` and overwrites all existing settings

Runtime settings (ie. env vars, system properties) should have priority over hardcoded settings (ie. `log4j.properties`).

This would for example allow setting the `environment` default value to dev in `log4j.properties`, and overwrite it in production using a system property.

**Comment by bretthoerner** [comment]
I agree this is wrong, I'll try to fix soon, thanks.

**Comment by bretthoerner** [comment]
@georgms Do you use `jul` or `log4j`? The code you discussed was specific to `jul` and could be changed, but `log4j` (version 1) appears to set parameters manually after the constructor is called, so I'm not sure that we have a way to do the right thing there...

I can't decide whether to allow different logging systems to have different behaviors here. I'm tempted to say that people should just not set values in the properties if they want to override them at runtime. :disappointed:

**Comment by georgms** [comment]
@bretthoerner I'm actually using `log4j`. Turns out I was looking in the wrong files...

But I've just checked again, and the described behavior exists with `log4j` as well. Here is what I've found (hopefully the right files this time):

* When the application is started up `log4j` will be initialized, and through `log4j.LogManager`'s static constructor the `SentryAppender` will be constructed
* `SentryAppender`'s constructor looks up the settings (eg. `setEnvironment(Lookup.lookup("environment"));`) and sets them
* Later on in `log4j.PropertyConfigurator#parseAppender` the settings from `log4j.properties` for the `SentryAppender` are parsed and set in the `SentryAppender` instance

For reference, my `log4j.properties` looks about like this:

```properties
# default to log >=WARN messages to console, file and error-log
log4j.rootCategory=WARN, Console, File, Sentry

log4j.appender.Console=...
log4j.appender.File=...

log4j.appender.Sentry=com.getsentry.raven.log4j.SentryAppender
log4j.appender.Sentry.extraTags=x,y
log4j.appender.Sentry.ravenFactory=com.my-company.sentry.ReleaseAwareRavenFactory
log4j.appender.Sentry.dsn=...
log4j.appender.Sentry.environment=dev
log4j.appender.Sentry.Threshold=ERROR
```

Trying to change the environment by starting up my application with `-Dsentry.environment=prod` has no effect, all events will still be logged with `environment=dev`.

However, when I remove the `environment` line from `log4j.properties`, the `environment` system property works fine.

**Comment by bretthoerner** [comment]
> Later on in log4j.PropertyConfigurator#parseAppender the settings from log4j.properties for the SentryAppender are parsed and set in the SentryAppender instance

That's the issue for us, since this is out of our control and happens *after* we setup environment & Java property settings. Any ideas? I feel like we're stuck.

**Comment by bretthoerner** [comment]
I actually found a solution for this in #355.

**Comment by bretthoerner** [comment]
Fixed here: https://github.com/getsentry/raven-java/commit/ee2d750f18ccc56e47d05f545800947e2430bbb9

To be released in `8.0.0` which will be released in a ~week I hope.
